kani-compiler/src/kani_middle/transform/loop_contracts.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains code related to the MIR-to-MIR pass to enable loop contracts.
+//!
+
+use crate::kani_middle::codegen_units::CodegenUnit;
+use crate::kani_middle::find_fn_def;
+use crate::kani_middle::transform::body::{MutableBody, SourceInstruction};
+use crate::kani_middle::transform::{TransformPass, TransformationType};
+use crate::kani_middle::KaniAttributes;
+use crate::kani_queries::QueryDb;
+use crate::stable_mir::CrateDef;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Symbol;
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::{BasicBlockIdx, Body, Operand, Terminator, TerminatorKind};
+use stable_mir::ty::{FnDef, RigidTy};
+use stable_mir::DefId;
+use std::collections::VecDeque;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use tracing::trace;
+
+/// This pass will perform the following operations:
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
+///    to invoke the closure.
+///
+/// 2. Replace the dummy call to the register function with the actual call, i.e., transform
+///
+/// ```ignore
+/// let kani_loop_invariant = || -> bool {inv};
+/// kani_register_loop_contract(kani_loop_invariant)
+/// while guard {
+///     loop_body;
+///     kani_register_loop_contract(||->bool{true});
+/// }
+/// ```
+///
+///    to
+///
+/// ```ignore
+/// let kani_loop_invariant = || -> bool {inv};
+/// while guard {
+///     loop_body;
+///     kani_register_loop_contract(kani_loop_invariant);
```

Comment:
This could break the aliasing and ownership model.

Reply:
I understand that it will violate the borrow check in Rust. But is it also a problem if we do it in MIR level?

Reply:
Yes. The borrow checker is trying to prevent users from making mistakes since mutation may invalidate memory.

Here is an example where I think this won't work:

```rust
type Data = u8;

struct Container {
    data: Box<Data>,
}

impl Container {
    fn new_data(&mut self, data: Data) {
        self.data = Box::new(data)
    }
}

/// Dummy function that reallocates data inside the loop.
pub fn target() {
    let mut container = Container { data: Box::new(1) };
    let mut i = 0;
    // This captures the pointer to data that will be deallocated
    let invariant = || {
        *container.data > i
    };
    kani_register_loop_contract(invariant);
    while i < 10 {
        i += 1;
        container.new_data(*container.data + i);
        let place_holder = || { true };
        kani_register_loop_contract(place_holder);
    }
}

const fn kani_register_loop_contract<T, F: FnOnce() -> T>(f: F) -> T {
    unreachable!()
}
```

I believe the transformation from this PR would trigger an invalid read by looking at the MIR generated using [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=458da18d5b304ebcf38ddec1224a5884).

It looks like the `invariant` closure captures a pointer to the allocated data (not a reference to the box):
```asm
_1 = Container { data: move _2 };
_3 = const 0_u8;
_18 = deref_copy (_1.0: std::boxed::Box<u8>);
_20 = (((_18.0: std::ptr::Unique<u8>).0: std::ptr::NonNull<u8>).0: *const u8);
_5 = &(*_20);
_6 = &_3;
_4 = {closure@[src/lib.rs:18:21: 18:23](..)} { container: move _5, i: move _6 };
```

Thus, invoking this closure in the latch (after `new_data()` call that reallocates the box) will trigger an invalid memory read.

Can you please check if that's the case?

Reply:
Good point. I checked the case and it does cause problem in current transformation. I am thinking about not using closure to capture the invariants, and rolling back to the marker functions idea. I will try to refactor my code and update here.